pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -495,23 +493,38 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 		if identityMatches(set, replicas[i]) && storageMatches(set, replicas[i]) && retentionMatch {
 			continue
 		}
-		// Make a deep copy so we don't mutate the shared cache
-		replica := replicas[i].DeepCopy()
-		if err := ssc.podControl.UpdateStatefulPod(ctx, updateSet, replica); err != nil {
-			return &status, err
-		}
+		fns = append(fns, func() (bool, error) {
+			// Make a deep copy so we don't mutate the shared cache
+			replica := replicas[i].DeepCopy()
+			if err := ssc.podControl.UpdateStatefulPod(ctx, updateSet, replica); err != nil {
+				return true, err
+			}
+			return false, nil
+		})
 	}
 
+	if monotonic {
+		for _, fn := range fns {
+			if shouldExit, err := fn(); shouldExit || err != nil {
+				return &status, err
+			}
+		}
+	} // parallel will run later
```

Comment:
Or maybe we can run this here as well?

This will increase readability, as me (as a code reader), I won't need to remember that there are some actions to execute.

Also from performance perspective, it shouldn't matter much as we don't expect pods to be created and terminated at the same time.

Reply:
For me it's easier to reason about this logic if we gather all actions first and then execute them. I agree it probably won't impact performance much when scaling.